### ISSUE TITLE
firestore sdk unable to see server errors due to CORS

### ISSUE THREAD

**Issue by ribrdb** [firebase/firebase-js-sdk#6793]
### [REQUIRED] Describe your environment

  * Operating System version: macos ventura 13.0.1
  * Browser version: Google Chrome	107.0.5304.110 
  * Firebase SDK version: 9.9.0
  * Firebase Product: firestore

### [REQUIRED] Describe the problem

During the firestore incident on 2022-11-09, our app was experiencing a total outage although the incident report says only streaming reads should have been affected.

The chrome console said:
```
instrument.ts:124 [2022-11-09T15:01:01.315Z]  @firebase/firestore: Firestore (9.9.0): Could not reach 
Cloud Firestore backend. Backend didn't respond within 10 seconds.
This typically indicates that your device does not have a healthy Internet connection at the moment. 
The client will operate in offline mode until it is able to successfully connect to the backend.
(anonymous) @ instrument.ts:124

elided.com/:1 Access to fetch at 'https://firestore.googleapis.com/google.firestore.v1.Firestore/Listen/channel?database=...' 
from origin 'https://elided.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' 
header is present on the requested resource. If an opaque response serves your needs, set the request's 
mode to 'no-cors' to fetch the resource with CORS disabled.

firestore.googleapis.com/google.firestore.v1.Firestore/Listen/channel?database=...:1          Failed to load resource: 
net::ERR_FAILED
instrument.ts:124 [2022-11-09T15:01:30.517Z]  @firebase/firestore: Firestore (9.9.0): Connection WebChannel transport errored: ir
(anonymous) @ instrument.ts:124 
```

It appears the 500 errors returned by the firestore api were missing the Access-Control-Allow-Origin header. So the browser blocks the requests and the sdk assumes the user's network is offline. 
Is it possible for the server to add CORS headers even when returning a 500? Or is there some other way to tell if a missing response might be due to a server error instead of assuming it is always the user's network?

**Comment by ribrdb** [comment]
I forgot to link to the incident: https://status.cloud.google.com/incidents/w7D5XGtVfMGcdcg4Zzud

**Comment by Opelor** [comment]
I have a similar issue. It works fine with older projects. It fails with a new one.

I can login using the SDK but I can't write anything.

See the continuous ping/fetch 

![Screen Shot 2022-11-14 at 1 24 42 PM](https://user-images.githubusercontent.com/3394156/201737431-8ed3dbd8-d29d-4489-bc53-271cead7746a.png)

**Comment by ehsannas** [comment]
Thanks for reporting @ribrdb . I'll take a look.

**Comment by ehsannas** [comment]
Googlers see b/259438625

**Comment by ehsannas** [comment]
The logs show that the backend server was crashed and was unreachable.

**Comment by ribrdb** [comment]
Yes the backend was unreachable. But shouldn't the client treat this as a server error, not go into offline mode?

**Comment by ribrdb** [comment]
And some google server was returning an error, otherwise chrome wouldn't have blocked a request. It seems like the best solution would be that whatever server returned that error could add Access-Control-Allow-Origin to the error so the client can see the error. If not, you could use something like https://google.github.io/closure-library/api/goog.labs.net.webChannel.netUtils.html#testNetwork to test if the error is caused by the user's network or the server. That has the disadvantage that you have to wait 10 seconds for the request to timeout though.

**Comment by ehsannas** [comment]
This fix is internally tracked by b/261731455. Closing.
